src/mongo_types.ts

```diff
@@ -17,14 +17,16 @@ import type { Sort } from './sort';
 export type TODO_NODE_3286 = any;
 
 /** Given an object shaped type, return the type of the _id field or default to ObjectId @public */
-export type InferIdType<TSchema> = TSchema extends { _id: infer IdType } // user has defined a type for _id
-  ? // eslint-disable-next-line @typescript-eslint/ban-types
-    {} extends IdType // TODO(NODE-3285): Improve type readability
-    ? // eslint-disable-next-line @typescript-eslint/ban-types
-      Exclude<IdType, {}>
-    : unknown extends IdType
-    ? ObjectId
+export type InferIdType<TSchema> = TSchema extends { _id: infer IdType }
+  ? // user has defined a type for _id
+    Record<any, never> extends IdType
+    ? never // explicitly forbid empty objects as the type of _id
     : IdType
+  : TSchema extends { _id?: infer IdType }
+  ? // optional _id defined - return ObjectId | IdType
+    unknown extends IdType
+    ? ObjectId // infer the _id type as ObjectId if the type of _id is unknown
+    : IdType | ObjectId
```

Comment:
Are we covering the optional id case with this line? If so, I think we may still want to keep it as `IdType`, we're just relying on the user to make sure that's actually true via their own pkFactory definition. While it's technically more correct to report `IdType | ObjectId`, since if they *don't* specify the pkFactory, ObjectId will be generated, by the same logic, if their pkFactory returns `SomeOtherIdType`, there's no way for us to catch that. Thus, since it's impossible for us to really make any guarantees about the type in this case, we may as well take their word for the type and just faithfully stick to `IdType` (and not `| ObjectId`) for our inferred ID. But, I'm open to being convinced otherwise. @nbbeeken @baileympearson

Reply:
I could also go either way on this change.  I made this change to address NODE-3761 (as discussed during triage today).

One argument for keeping the union is that even though we're now recommending that `_id`s be optional, there are probably users out there not using the pkFactory but who have `_id` as optional on their schemas.  Adding the correct return type here would potentially steer users away from using optional `_id`s by making it slightly more cumbersome to use (since any time they attempt to access the `_id`, they'll have to deal with both type cases).  

I would say we should decide which of these scenarios would be more important for us

- making the typing straightforward for users of pkFactory and not using the more correct type definition
- using the correct type definition to try and steer users away from optional `_id` fields and in the process making it a bit more cumbersome for pkFactory users

I kind of lean towards option 1 but I'd like to hear what @nbbeeken  thinks too

Reply:
Basically, the idea os that users shouldn't use optional _id unless they have a pkFactory method enforcing it; for users who want to optionally sometimes specify an ObjectId on their inserts, they are already covered by the "omit _id from your schema" option.

Reply:
Okay, that works.  I can remove the `| ObjectId`